File: 0214-shortest-palindrome/0214-shortest-palindrome.java
class Solution {
    private void computeLPS(String pattern, int[] lps) {
        int M = pattern.length();
        int len = 0;
        lps[0] = 0;
        int i = 1;

        while (i < M) {
            if (pattern.charAt(i) == pattern.charAt(len)) {
                len++;
                lps[i] = len;
                i++;
            } else {
                if (len != 0) {
                    len = lps[len - 1];
                } else {
                    lps[i] = 0;
                    i++;
                }
            }
        }
    }
    public String shortestPalindrome(String s) {
        String rev = new StringBuilder(s).reverse().toString();
        String temp = s + "#" + rev;

        int[] lps = new int[temp.length()];
        computeLPS(temp, lps);

        int longestPalindromePrefixLen = lps[lps.length - 1];

        String ans = rev.substring(0, s.length() - longestPalindromePrefixLen) + s;
        return ans;
    }
}





// class Solution {
//     public String shortestPalindrome(String s) {
//         String rev = new StringBuilder(s).reverse().toString();
//         int n = s.length();
//         for (int i = 0; i < n; i++) {
//             if (s.substring(0, n - i).equals(rev.substring(i))) {
//                 return rev.substring(0, i) + s;
//             }
//         }
//         return rev + s;
//     }
// }

File: LeetCode/0214-shortest-palindrome/0214-shortest-palindrome.py
class Solution:
    def computeLPS(self, pattern: str) -> list[int]:
        M = len(pattern)
        lps = [0] * M
        length = 0
        i = 1

        while i < M:
            if pattern[i] == pattern[length]:
                length += 1
                lps[i] = length
                i += 1
            else:
                if length != 0:
                    length = lps[length - 1]
                else:
                    lps[i] = 0
                    i += 1
        return lps

    def shortestPalindrome(self, s: str) -> str:
        rev = s[::-1]
        temp = s + "#" + rev

        lps = self.computeLPS(temp)

        longest_palindrome_prefix_len = lps[-1]

        ans = rev[:len(s) - longest_palindrome_prefix_len] + s
        return ans





# class Solution:
#     def shortestPalindrome(self, s: str) -> str:
#         rev = s[::-1]
#         n = len(s)
#         for i in range(n):
#             if s[:n - i] == rev[i:]:
#                 return rev[:i] + s
#         return rev + s